File: problems/931-Minimum-Falling-Path-Sum/solution.go
func minFallingPathSum(matrix [][]int) int {
	row, col := len(matrix), len(matrix[0])

	dp := make([][]int, row)
	for i := 0; i < col; i++ {
		dp[i] = make([]int, col)
	}

	for i := 0; i < col; i++ {
		dp[row-1][i] = matrix[row-1][i]
	}

	for i := row - 2; i >= 0; i-- {
		for j := 0; j < col; j++ {
			if j == 0 {
				if dp[i+1][j] <= dp[i+1][j+1] {
					dp[i][j] = matrix[i][j] + dp[i+1][j]
				} else {
					dp[i][j] = matrix[i][j] + dp[i+1][j+1]
				}
			} else if j == col-1 {
				if dp[i+1][j-1] <= dp[i+1][j] {
					dp[i][j] = matrix[i][j] + dp[i+1][j-1]
				} else {
					dp[i][j] = matrix[i][j] + dp[i+1][j]
				}
			} else {
				if dp[i+1][j-1] <= dp[i+1][j] && dp[i+1][j-1] <= dp[i+1][j+1] {
					dp[i][j] = matrix[i][j] + dp[i+1][j-1]
				} else if dp[i+1][j] <= dp[i+1][j-1] && dp[i+1][j] <= dp[i+1][j+1] {
					dp[i][j] = matrix[i][j] + dp[i+1][j]
				} else {
					dp[i][j] = matrix[i][j] + dp[i+1][j+1]
				}
			}
		}
	}

	min := 1000000
	for i := 0; i < col; i++ {
		if dp[0][i] <= min {
			min = dp[0][i]
		}
	}

	return min
}

File: problems/931-Minimum-Falling-Path-Sum/solution.java
class Solution {
    public int minFallingPathSum(int[][] matrix) {
        if (matrix.length == 1) {
            return matrix[0][0];
        }

        for (int i = 1; i < matrix.length; i++) {
            matrix[i][0] = Math.min(matrix[i-1][0], matrix[i-1][1]) + matrix[i][0];
            
            for (int j = 1; j < matrix[i].length-1; j++) {
                matrix[i][j] = Math.min(Math.min(matrix[i-1][j-1], matrix[i-1][j+1]), matrix[i-1][j]) + matrix[i][j];
            }

            matrix[i][matrix[i].length-1] = Math.min(matrix[i-1][matrix[i-1].length-2], matrix[i-1][matrix[i-1].length-1]) + matrix[i][matrix[i].length-1];
        }

        int result = Integer.MAX_VALUE;
        for (int i = 0; i < matrix[matrix.length-1].length; i++) {
            result = Math.min(result, matrix[matrix.length-1][i]);
        }

        return result;
    }
}

